### PR TITLE
Implement skip list for version compatibility lint step

### PR DESCRIPTION
**Pull request by ebresafegaga**
Implements the skip list for #11780, and closes it. This will help to unblock #11558.

### CODE REVIEW

**Comment by psteckler** [comment]
There's a plan for the version linter, not yet implemented, which will get rid of most of the linter noise. The `compare_versioned_types` scripts is going away.

**Comment by ebresafegaga** [comment]
> There's a plan for the version linter, not yet implemented, which will get rid of most of the linter noise. The `compare_versioned_types` scripts is going away.

But until then, maybe this could be a temporary fix to avoid blockers with some new PRs?

**Comment by kantp** [comment]
> There's a plan for the version linter, not yet implemented, which will get rid of most of the linter noise. The `compare_versioned_types` scripts is going away.

Thanks, that's great to know! We should make a decision for what to do in the meantime, until we have the new version linter. I would advocate for implementing the skip list, to avoid blocking PRs like #11558 and #11629. Alternatively, we could also disable the check until we have the new version linter, but I think that would be going too far.

**Comment by psteckler** [comment]
I'm not sure I understand the approach here.

A type doesn't pass or fail the linter in isolation. A type fails the linter if it's different in the PR branch from the same type in the PR base branch.

You might not want to issue a warning if the type is not in use on mainnet. In that situation, users are not depending on the serialization of that type.  But the current setup doesn't allow use to identify which types are in use on mainnet, so the linter gives undesired noise. The upcoming changes will do that, by checking against a "release" branch.

When such noisy warnings occur, we've been manually merging PRs, after examining the contents of the linter warning. For example, changes to zkApps-related types, which are not in use on mainnet, can be merged despite linter warnings. It happens, but not super often.

I'm also worried that the approach here is too coarse-grained: a file might contain types that you want to lint, as well as types that you don't want to lint.

**Comment by kantp** [comment]
Ok, let me explain the reasoning behind this proposed change: to my understanding, there are currently some files that do not pass the linter, which have not been detected, because the linter is only run when files are changed. This causes CI to fail on PRs if they make some change within such a file (as in #11558). Is that understanding correct?

If it is, the ideal solution would be run the linter on all files, and edit those that fail to pass the linter. But that would be a lot of work, and it's not clear to me that it's a thing we need to do right now (but maybe there is?). So, as a stopgap, the skip list would allow us to say "this file doesn't pass the linter, but it never did, so CI should not block PRs modifying the file", and remove the blockage.

Alternatively, I would also be fine with force merging PRs such as #11558 (though I would still like the error message in the linter to refer to this possibility, so that people are not surprised by CI failing on something that is unrelated to their PR).

File: scripts/compare_versioned_types.py [hunk 1]
```diff
@@ -10,10 +10,27 @@
...
 from compare_versioned_items import run_comparison
 
+error_msg = "Please see this issue https://github.com/MinaProtocol/mina/issues/11780 for pointers on how to fix this error. It might be also helpful to look in this file: `scripts/compare_verioned_types.py"
+
+# This is the list of file while has never been changed before
+# and they haven't run through the CI before, and then fail 
+# when modified. To add new files, just add them to this list.
+skip_list = ["src/nonconsensus/snark_params/tick.ml"]
+
+def skip(original, modified):
```
Comment (joseandro) [inline scripts/compare_versioned_types.py:20]: Quoting the above comment here:
> This is the list of file while has never been changed before
> and they haven't run through the CI before, and then fail when modified.

Some observations:
1. The `modified` parameter isn't being used in the `skip` function.
2. How are you accounting for failures when the skip list is modified?
3. Can you add some documentation to the `skip` function? It's not clear to me what both parameters mean.

**Review by joseandro — CHANGES_REQUESTED** [review]
Left some comments in this PR requesting changes.

**Comment by bkase** [comment]
If we (a) make this file [scripts/compare_versioned_types.py](https://github.com/MinaProtocol/mina/pull/11789/files#diff-a7955128b7a2cf1faf8d02302417c189649253a87b184223d92757a2d05e2d56) have a strict CODEOWNERS so we don't accidentally add bad files (delegate to @joseandro for deciding which person(s) or teams should be on that) and (b) only add files that we haven't touched since adding the linter (which maybe we can add in a comment above the list), then this seems safe to me

File: scripts/compare_versioned_types.py (hunk 1 above)
Comment (ebresafegaga) [reply scripts/compare_versioned_types.py:20]: The `skip` function determines if a particular file should be skipped  version compatibility check in the CI check.
`modified` is the name of the file we have modified in the branch of a PR.
`original` is the name of the equivalent file in at the HEAD of the branch you want to merge into (this is usually develop or compatible).
They basically have the same names, but different paths. I'll add these as documentation to the file.

> 2. How are you accounting for failures when the skip list is modified?

What do you mean by this?

**Comment by bkase** [comment]
The comments look good to me but we need the CODEOWNERS change too, I’ll ask @joseandro to comment again

File: scripts/compare_versioned_types.py (hunk 1 above)
Comment (joseandro) [reply scripts/compare_versioned_types.py]: Do you need `modified` variable at all? I don't see it being used anywhere inside the function.

**Review by joseandro — APPROVED** [review]

**Comment by kantp** [comment]
!ci-build-me

**Comment by psteckler** [comment]
The updated version linter task is described in detail here: 

https://github.com/MinaProtocol/mina/issues/11181#issuecomment-1278106386

It might be worth starting on that, so this PR is not needed.

**Comment by mrmr1993** [comment]
The scripts that this modifies have been removed and replaced with the approach described in #11181. Closing.

**Pull request closed without merging**